src/nvim/ex_getln.c

```diff
@@ -2131,6 +2131,27 @@ static int command_line_handle_key(CommandLineState *s)
     break;
   }
 
+  char buf[MB_MAXBYTES + 1];
+  buf[utf_char2bytes(s->c, buf)] = NUL;
+
+  textlock++;
+  set_vim_var_string(VV_CHAR, buf, -1);
+  char firstcbuf[2];
+  firstcbuf[0] = (char)(s->firstc > 0 ? s->firstc : '-');
+  firstcbuf[1] = 0;
+
+  if (apply_autocmds(EVENT_CMDLINECHARPRE, firstcbuf, firstcbuf, false, curbuf)) {
+    if (strcmp(buf, get_vim_var_str(VV_CHAR)) != 0) {
+      textlock--;
+      put_on_cmdline(get_vim_var_str(VV_CHAR), -1, true);
+      set_vim_var_string(VV_CHAR, NULL, -1);
+      return command_line_changed(s);
+    }
+  }
+
+  set_vim_var_string(VV_CHAR, NULL, -1);
+  textlock--;
```

Comment:
This function is already too big. Can this be factored into a function?

Reply:
Yes it can.